### PR TITLE
Reduce ASM dep in `ClassDescriptor`

### PR DESCRIPTION
**Pull request by basil**
See [stapler/stapler#226 (comment)](https://github.com/stapler/stapler/pull/226#issuecomment-847359651). Amends that PR to try `MethodParameters` first (the most efficient and supported system) then fall back to using ASM for `LocalVariableTable` if necessary. Tested interactively in Jenkins core in the following scenarios:

- Stapler compiled with metadata for reflection on method parameters, Jenkins core compiled _without_ metadata for reflection on method parameters
- Both Stapler and Jenkins core compiled with metadata for reflection on method parameters via jenkinsci/pom#171

In the first scenario I verified that ASM was correctly used as a fallback. In the second scenario I verified that ASM was not used as a fallback.

### CODE REVIEW

File: core/src/test/java/org/kohsuke/stapler/ClassDescriptorTest.java [hunk 1]
```diff
@@ -28,6 +28,32 @@
...
+        expected.put("NoParams_static", new String[0]);
+        expected.put("ManyParams", new String[] { "a", "b", "c", "d", "e", "f", "g", "h", "i" });
+        expected.put("Params_static", new String[] { "abc", "def", "ghi" });
+        // run tests
+        for (Map.Entry<String,String[]> entry : expected.entrySet()) {
+            Method testMethod = testCases.get(entry.getKey());
+            assertNotNull("Method missing for " + entry.getKey(), testMethod);
+            String[] result = ClassDescriptor.loadParameterNamesFromReflection(testMethod);
+            assertNotNull("Null result for " + entry.getKey(), result);
+            if (!Arrays.equals(entry.getValue(), result)) {
```
Comment (jglick) [inline core/src/test/java/org/kohsuke/stapler/ClassDescriptorTest.java:49]: Could use a standard `Matcher` for this I think.

File: pom.xml [hunk 2]
```diff
@@ -60,6 +60,9 @@
     <changelist>999999-SNAPSHOT</changelist>
     <scmTag>HEAD</scmTag>
     <spotbugs.excludeFilterFile>${project.basedir}/../src/spotbugs/spotbugs-excludes.xml</spotbugs.excludeFilterFile>
+
+    <!-- Generate metadata for reflection on method parameters -->
+    <maven.compiler.parameters>true</maven.compiler.parameters>
```
Comment (jglick) [inline pom.xml:65]: Better style to use plugin `configuration`.

**Review by jglick — APPROVED** [review]

**Comment by jglick** [comment]
LMK if you intend to do any follow-ups; otherwise this can be merged as is.

File: pom.xml (hunk 2 above)
Comment (basil) [reply pom.xml:65]: > Better style

Do you have a reference to a style guide?

File: core/src/test/java/org/kohsuke/stapler/ClassDescriptorTest.java (hunk 1 above)
Comment (basil) [reply core/src/test/java/org/kohsuke/stapler/ClassDescriptorTest.java:49]: I was just trying to change the existing code as little as possible. This is pretty much a copy of `loadParametersFromAsm`. There is a lot that can be cleaned up in this repository. Once the dependencies are up-to-date I'll start looking into cleanups like this.

File: pom.xml (hunk 2 above)
Comment (jglick) [reply pom.xml:65]: Could not find one. I wrote up what I know: https://stackoverflow.com/a/67794133/12916

Comment (basil) [reply pom.xml]: OK, this still seems like a fairly subjective argument to me, but I trust your judgment.

File: core/src/test/java/org/kohsuke/stapler/ClassDescriptorTest.java (hunk 1 above)
Comment (jglick) [reply core/src/test/java/org/kohsuke/stapler/ClassDescriptorTest.java:49]: In particular the use of reflection in the ASM test is unnecessary (fixed in #226).

File: pom.xml [hunk 3]
```diff
@@ -89,6 +89,21 @@
...
+        <plugin>
+          <groupId>org.apache.maven.plugins</groupId>
+          <artifactId>maven-compiler-plugin</artifactId>
+          <executions>
+            <execution>
+              <goals>
+                <goal>compile</goal>
+              </goals>
+            </execution>
+          </executions>
```
Comment (jglick) [inline pom.xml:101]: No need to override this aspect of the superpom.

```suggestion
```

Comment (basil) [reply pom.xml]: Thanks for the suggestion. It should be obvious that I have very little knowledge, let alone understanding, of Maven.

File: pom.xml [hunk 4]
```diff
@@ -89,6 +89,14 @@
             </ignores>
           </configuration>
         </plugin>
+        <plugin>
+          <groupId>org.apache.maven.plugins</groupId>
```
Comment (jglick) [inline pom.xml:93]: By the way, for official plugins this can be omitted if you like:
```suggestion
```
or kept for legibility.

**Review by jglick — APPROVED** [review]

File: pom.xml [hunk 5]
```diff
@@ -89,6 +89,14 @@
...
+          <artifactId>maven-compiler-plugin</artifactId>
+          <configuration>
+            <!-- Generate metadata for reflection on method parameters -->
+            <parameters>true</parameters>
+          </configuration>
+        </plugin>
```
Comment (jglick) [inline pom.xml:99]:
```suggestion
```

after #236

**Comment by jglick** [comment]
Hmm https://github.com/stapler/stapler/pull/233/checks?check_run_id=2721994088 did not work?

**Pull request merged**